src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -352,12 +378,36 @@ public override AppendPrependIterator<TSource> Prepend(TSource item)
                 return new AppendPrependN<TSource>(_source, _prepended != null ? _prepended.Add(item) : new SingleLinkedNode<TSource>(item), _appended);
             }
 
+            private TSource[] LazyToArray()
+            {
+                Debug.Assert(GetCount(onlyIfCheap: true) == -1);
+
+                var builder = new LargeArrayBuilder<TSource>(initialize: true);
+
+                for (SingleLinkedNode<TSource> node = _prepended; node != null; node = node.Linked)
+                {
+                    builder.Add(node.Item);
+                }
+
+                builder.AddRange(_source);
+
+                if (_appended != null)
+                {
+                    foreach (TSource item in _appended.ToArray())
```

Comment:
@stephentoub `LargeArrayBuilder` does not have an indexer, since it cannot be implemented efficiently. (I did it before using some bitshifting tricks/log2, but it was too slow.) In addition, we don't know if there's even enough room in the buffers to hold the appended items-- for example, we could have added 40 items, meaning there's 64 slots, and we need to append 25 items.